src/base/static/components/place-detail-view/place-detail-survey.js

```diff
@@ -0,0 +1,192 @@
+// TODO: localization in this component.
+
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import {
+  Map as ImmutableMap,
+  OrderedMap as ImmutableOrderedMap,
+} from "immutable";
+import emitter from "../utils/emitter";
+
+import FormField from "../form-field";
+import SecondaryButton from "../ui-elements/secondary-button";
+import PlaceDetailSurveyResponse from "./place-detail-survey-response";
+import WarningMessagesContainer from "../ui-elements/warning-messages-container";
+import Avatar from "../ui-elements/avatar";
+
+import constants from "../constants";
+import { placeDetailSurvey as messages } from "../messages";
+
+import "./place-detail-survey.scss";
+
+class PlaceDetailSurvey extends Component {
+  constructor(props) {
+    super(props);
+    let fields = ImmutableOrderedMap();
+    this.props.surveyConfig.items.forEach(field => {
+      fields = fields.set(
+        field.name,
+        ImmutableMap()
+          .set(constants.FIELD_STATE_RENDER_KEY, Math.random())
+          .set(constants.FIELD_STATE_VALUE_KEY, "")
+      );
+    });
+    this.state = {
+      fields: fields,
+      isFormSubmitting: false,
+      formValidationErrors: new Set(),
+      showValidityStatus: false,
+    };
+  }
+
+  componentWillMount() {
+    emitter.addListener("place-detail-survey:save", () => {
+      let fields = ImmutableOrderedMap();
+      this.props.surveyConfig.items.forEach(field => {
+        fields = fields.set(
+          field.name,
+          ImmutableMap()
+            .set(constants.FIELD_STATE_RENDER_KEY, Math.random())
+            .set(constants.FIELD_STATE_VALUE_KEY, "")
+        );
+      });
+
+      this.setState({
+        fields: fields,
+        isFormSubmitting: false,
+        formValidationErrors: new Set(),
+        showValidityStatus: false,
+      });
+    });
+  }
+
+  componentWillUnmout() {
+    emitter.removeAllListeners("place-detail-survey:save");
+  }
+
+  onFieldChange(fieldName, fieldStatus, isInitializing) {
+    fieldStatus = fieldStatus.set(
+      constants.FIELD_STATE_RENDER_KEY,
+      this.state.fields.get(fieldName).get(constants.FIELD_STATE_RENDER_KEY)
+    );
+    this.setState(({ fields }) => ({
+      fields: fields.set(fieldName, fieldStatus),
+      updatingField: fieldName,
+      isInitializing: isInitializing,
+    }));
+  }
+
+  onSubmit(evt) {
+    evt.preventDefault();
+
+    const newValidationErrors = new Set();
+    let isValid = true;
+    this.state.fields.forEach(value => {
+      if (!value.get(constants.FIELD_STATE_VALIDITY_KEY)) {
+        newValidationErrors.add(
+          value.get(constants.FIELD_STATE_VALIDITY_MESSAGE_KEY)
+        );
+        isValid = false;
+      }
+    });
```

Comment:
I think this block would be simpler if you assigned `newValidationErrors` to be the result of a `filter` then `reduce` on `this.state.fields`. Then you can set `isValid` to be whether `newValidationErrors.size === 0`.